Introduce PEP for pip provenance

fridex opened the pull request.

hugovk committed:
Infra: Use consistent 'make html' and 'make dirhtml' (#2968)

methane committed:
PEP 582: Fix footnote (#2970)

ericsnowcurrently committed:
PEP 554: Simplify! (#2812)

Most notably, I've removed "channels" from the PEP.

hugovk committed:
Infra: replace feedgen/lxml dependency to test Python 3.12-dev (#2973)

hugovk committed:
Infra: Use relative links to ease development (#2972)

JelleZijlstra committed:
PEP 702: Changes after discussion (#2975)

bxparks committed:
PEP 495: Fix incorrect gap image (#2965)

The SVG image in the "Mind the Gap" section (pep-0495-gap.svg) was incorrect.
It is essentially the same as the the SVG image in the "In the Gap" section (pep-0495-fold.svg).
This is a bug that dates back to the original commit on
2015-08-30 (814d3728115b94fbede89f5c73f5cf3065d00e7b)
that uploaded the both the PNG and SVG versions of the Fold and Gap images.
The PNG version of the Gap image is correct, but the SVG version is incorrect.

The replacement SVG image attached in this PR is a manual recreation
of the PNG version of the Gap image that was deleted on 2022-04-23.
It is not exactly the same, but I think it is close enough.

JelleZijlstra committed:
PEP 702: Add to Post-History (#2976)

colesbury committed:
PEP 703: --disable-gil and other small changes (#2979)

* Rename `--without-gil` to `--disable-gil`.

Suggested by Inada Naoki.

* Add two sections to "Rejected Ideas":

 - Why Not Deprecate ``PyDict_GetItem`` in Favor of ``PyDict_FetchItem``?
 - Why Not Use PEP 683 Immortalization?

* Expand on Python build modes

hugovk committed:
PEP 13: Update current SC members for 2023 term (#2978)

PEP 13: Update current SC members

hugovk committed:
Infra: Fix Replaces/Superseded-By/Requires links for dirhtml (#2981)

After #2972, when building dirhtml,
we need to re-add the initial / for the links to PEPs
in the Replaces/Superseded-By/Requires headers.

hugovk committed:
Infra: Use dark grey instead of dark green for dark theme background (#2977)

* Use #121212 for dark theme bgcolor

* Use #1111111 for dark theme bgcolor

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

JelleZijlstra committed:
PEP 702: Fix typo, section ordering (#2982)

pradyunsg committed:
PEP 704: Require virtual environments by default for installers (#2964)

Enforce the use of virtual environments and establish a conventional
name for the virtual environment directory.

Co-authored-by: Pradyun Gedam <pradyunsg@users.noreply.github.com>
Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>
Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>
Co-authored-by: Zak Johnson <me@zakj.net>

Commit:
PEP 705: Recording provenance of installed packages

Signed-off-by: Fridolin Pokorny <fridolin.pokorny@datadoghq.com>

fridex commented on pep-9999.rst:
```diff
@@ -0,0 +1,89 @@
+
+This PEP describes a way to record provenance of Python packages installed.
+This record is created by an installer and is available to users in a form of a
+JSON file ``direct_url.json`` in ``.dist-info`` directory. This PEP extends
```
Maybe it would worth to have this renamed:

```suggestion
JSON file ``provenance.json`` in ``.dist-info`` directory. This PEP extends
```

trishankatdatadog commented on pep-9999.rst:
```diff
@@ -0,0 +1,89 @@
+The JSON document is stating the following entries:
+
+* ``archive_info.hash`` MUST be present with a value of ``<hash-algorithm>=<expected-hash>``,
+  currently supported hash algorithm is only ``sha256``.
```
Could we also add `blake2b-256`? That is what [PEP 458](https://github.com/pypi/warehouse/pull/10870/files) plans to use after discussion with PyPA folks.

fridex replied:
Sounds reasonable, I'm not sure how the ``archive_info.hash`` should look like in such a case as ``<hash-algorithm>`` is directly part of its value (maybe comma separated?). This might also open discussion how it should look like in already existing support for ``direct_url.json``.

trishankatdatadog commented on pep-9999.rst:
```diff
@@ -0,0 +1,89 @@
+PEP: 9999
+Title: Recording provenance of installed packages
```
On second thought, maybe we shouldn't call it "provenance", which is a bit overloaded. Maybe "installation reports" is better. Better yet, maybe it could even be a new type of [in-toto attestation](https://github.com/in-toto/attestation).

fridex replied:
> On second thought, maybe we shouldn't call it "provenance", which is a bit overloaded. Maybe "installation reports" is better. 

Installation report can be confused with [pip's installation reports](https://pip.pypa.io/en/stable/reference/installation-report/).

> Better yet, maybe it could even be a new type of [in-toto attestation](https://github.com/in-toto/attestation).

Maybe we could add this directly to PEP as an example of usage, beside SBOM already mentioned.

trishankatdatadog commented on pep-9999.rst:
```diff
@@ -0,0 +1,89 @@
+`PEP-610 <https://peps.python.org/pep-0610/>`_.
+
+
+Motivation
```
One important thing from our discussion: what does this proposal give you that installation reports already don't? One may argue that installation reports are too verbose and too specific to the OS/Python versions. However, I think the more significant philosophical difference is that you want these reports/provenance to be produced by default, since most end-users are not going to go through the trouble to explicitly produce them unless compelled. However, even if these reports are produced by default, the problem remains that old applications/containers are going to be missing these reports.

The other issue is whether something higher-level like Hatch might be better suited to producing this information, because then you can associate _applications_ with packages.

Fidget-Spinner committed:
PEP 699: Add implementation link (#2984)

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>

pablogsal committed:
PEP 701: Incorporate more feedback from the discussion thread (#2974)

kushaldas committed:
PEP 582: Add more explanation based on feedback (#2983)

* Updates based on feedback

Adds explanation of handling `bin/` directory, or how we are not
trying to replace the existing virtual environments.

* Updates grammar based on feedback

Commit:
Apply suggestions from code review

Co-authored-by: C.A.M. Gerlach <CAM.Gerlach@Gerlach.CAM>
Co-authored-by: Hugo van Kemenade <hugovk@users.noreply.github.com>

fridex commented:
Closing, replaced with https://github.com/fridex/peps/pull/2.
